lib/js/src/manager/screen/BaseTextAndGraphicManager.js

```diff
@@ -0,0 +1,956 @@
+                } else {
+                    currentWindowID = PredefinedWindows.DEFAULT_WINDOW;
+                }
+                if (currentWindowID === PredefinedWindows.DEFAULT_WINDOW) {
+                    this._defaultMainWindowCapability = windowCapability;
+                }
+            }
+        };
+
+        this._lifecycleManager.addRpcListener(FunctionID.OnHMIStatus, this._hmiListener);
+        this._lifecycleManager.addOnSystemCapabilityListener(SystemCapabilityType.DISPLAYS, this._onDisplayCapabilityListener);
+    }
+
+    /**
+     * Conditionally applies a Show update based on what has been set in the internal state
+     * @param {Boolean} - Whether it's known that there's a change that warrants an update
+     * @return {Promise}
+    */
+    async _update (knownUpdate) {
+        if (knownUpdate) {
+            this._taskQueue.push(true);
+        }
+        // don't continue if the manager is in batch mode
+        if (this._batchingUpdates) {
+            return;
+        }
+        if (!this._isUpdateRunning) {
+            this._isUpdateRunning = true;
+            await this._sdlUpdate();
+            this._isUpdateRunning = false;
+        }
+    }
+
+    /**
+     * Determines what needs to be done to send a valid Show method
+     * @return {Promise}
+    */
+    async _sdlUpdate () {
+        // don't continue if there's nothing to do or if the manager didn't get the right HMI level yet
+        if (!this._isReady || this._taskQueue.length === 0) {
+            return;
+        }
+        // empty the queue out the safe way
+        while (this._taskQueue.length > 0) {
+            this._taskQueue.shift();
+        }
+
+        // Updating Text and Graphics
+        let fullShow = new Show();
+        if (this._textAlignment !== null) {
+            fullShow.setAlignment(this._textAlignment);
+        }
+
+
+        fullShow = this._assembleShowImages(this._assembleShowText(fullShow));
+
+        if (!this._shouldUpdatePrimaryImage() && !this._shouldUpdateSecondaryImage()) {
+            // No Images to send, only sending text
+            await this._sendShow(this._extractTextFromShow(fullShow));
+        } else if (!this._sdlArtworkNeedsUpload(this._primaryGraphic) &&
+            (this._secondaryGraphic === this._blankArtwork || !this._sdlArtworkNeedsUpload(this._secondaryGraphic))) {
+            // Images already uploaded, sending full update
+            // The files to be updated are already uploaded, send the full show immediately
+            await this._sendShow(fullShow);
+        } else {
+            // Images need to be uploaded, sending text and uploading images
+            const success = await this._uploadImages();
+            if (!success) {
+                fullShow = this._extractTextFromShow(fullShow);
+            }
+            await this._sendShow(fullShow);
+        }
+
+        // run this method again for potential future tasks
+        return this._sdlUpdate();
```

Comment:
Java Suite's comparable method doesn't seem to call itself recursively, but perhaps this is the intended behavior?